### ISSUE TITLE
Additional Kotlin DSL Overloads for TaskContainer.register(ing)

### ISSUE THREAD

**Issue by Fleshgrinder** [gradle/gradle#16915]
I am always trying to follow the best practices, and one of those (even if only mentioned in the community Slack) is to name a task class exactly like the corresponding task should be named. Consequently it is an annoyance that I have to repeat the name whenever I use `register` anywhere, instead of simply having it automatically set up for me:

```kotlin
public inline fun <reified T : Task> TaskContainer.register(): TaskProvider<T> =
    register(T::class.java.simpleName.decapitalize(Locale.ROOT), T::class.java)

val t = tasks.register<MyTask>()
assert(t.name == "myTask")
```

For `registering` it would be nice to have the ability, as we already have it for `register`, to pass the type via a type parameter, instead of a value parameter. I know that `registering` is actually provided through another less concrete super type, but since the `TaskContainer` is a type that pretty much every user touches while using Gradle it's imho worth it to provide additional overloads that work like they are used to from many other Kotlin APIs.

```kotlin
public inline fun <reified T : Task> TaskContainer.registering(): ReadOnlyProperty<Any?, TaskProvider<T>> =
    ReadOnlyProperty<Any?, TaskProvider<T>> { _, property -> register(property.name, T::class.java) }

public inline fun <reified T : Task> TaskContainer.registering(configure: Action<T>): ReadOnlyProperty<Any?, TaskProvider<T>> =
    ReadOnlyProperty { _, property -> register(property.name, T::class.java, configure) }

public inline fun <reified T : Task> TaskContainer.registering(vararg args: Any): ReadOnlyProperty<Any?, TaskProvider<T>> =
    ReadOnlyProperty { _, property -> register(property.name, T::class.java, args) }

val myTask by tasks.registering<MyTask>()

val myTask by tasks.registering<MyTask> { assert(this is MyTask) }

val myTask by tasks.registering<MyTask>("arg", "arg", "arg")
```

Of course I'm willing to provide the implementation with tests in a PR if this change is deemed useful.

**Comment by vlsi** [comment]
```kotlin
val myTask by tasks.registering<MyTask> {
  ...
}
```

 ^^^ That would be great

----

> I have to repeat the name whenever I use register anywhere, instead of simply having it automatically set up for me

I guess if you use `val myTask by tasks.registering<MyTask>()`, then it would be readable, and source code search would show the places where the task is created and configured.

**Comment by Fleshgrinder** [comment]
> I guess if you use `val myTask by tasks.registering<MyTask>()`, then it would be readable, and source code search would show the places where the task is created and configured.

That's what I use in build scripts, but while developing plugins and while I might rename it multiple times it's annoying that renaming the type alone is not enough to change the name of the default task that I register in the plugin. There I'd prefer to use `val task = project.tasks.register<MyTask>()` and move on knowing that it will be called `myTask`, and when I refactor to `YourTask` that it's named `yourTask`.

**Comment by vlsi** [comment]
> I might rename it multiple times it's annoying that renaming the type alone is not enough to change the name of the default task that I register in the plugin

Both task name and task type are public API of the plugin, so you should not change them often.
In other words, your end users might hard-code task names, and they might configure task via API (which statically compiles the class name).

If you have a fixed name for the task, then it is a nice idea to expose that constant via `public API` like in `LifecycleBasePlugin.CLEAN_TASK_NAME`.
Once you have the constant for `CLEAN_TASK_NAME`, you no longer need "automatic decapitalization" as you use the very same constant when registering the task like in

```java
final Provider<Delete> clean = project.getTasks().register(CLEAN_TASK_NAME, Delete.class, cleanTask -> {
        cleanTask.setDescription("Deletes the build directory.");
        cleanTask.setGroup(BUILD_GROUP);
        cleanTask.delete(buildDir);
});
```

On top of that, `decapitalize` is deprecated for removal: https://kotlinlang.org/docs/whatsnew1430.html#locale-agnostic-api-for-upper-lowercasing-text

In other words, I think I understand your demand, however, I believe it is better to refrain from adding that decapitalization as it results in doubtful practices.

```kotlin
val task = project.tasks.register<MyTask>()
```

`myTask` is a bad name for the task as it is obscure what it does. Would you please use a real-life-looking name for your example?
Sometimes I happen to add `...Task` to the class names for readability, however, I do not want my task names to end with `Task`.

**Comment by Fleshgrinder** [comment]
I know that decapitalize is deprecated, I contributed to the proposal and hoped that it would land. Too many mistakes where made by people who forgot to add `Locale.ROOT` to their calls. Let's hope life will be better for all of them in the future. 😊 The code in the PR description is example code to show how it could be done. I'd provide a final implementation in a PR with tests and everything.

> Both task name and task type are public API of the plugin, so you should not change them often.
> In other words, your end users might hard-code task names, and they might configure task via API (which statically compiles the class name).

Exactly, this is why we can use the class name directly for the task because renaming the class is as much a breaking change as renaming the corresponding task.

> If you have a fixed name for the task, then it is a nice idea to expose that constant via `public API` like in `LifecycleBasePlugin.CLEAN_TASK_NAME` so your users could use the constant instead of hard-coding the task name.

I see little value in this for end users, because of the code that is automatically generated and other nicer APIs to achieve the same, e.g.:

```kotlin
// build.gradle.kts

tasks.clean { doLast { println("it's all clean now") } }

val clean by tasks.existing

tasks.withType<Clean>().configureEach { /* ... */ }
```

> `myTask` is a bad name for the task as it is obscure what it does. Would you please use a real-life-looking name for your example?
> Sometimes I happen to add `...Task` to the class names for readability, however, I do not want my task names to end with `Task`.

As I wrote in the introduction, the whole thing would only be useful for people who follow the best practices, and that means that your task is named 1:1 like the task that is going to be visible to the users; examples:

- `class Publish : Task` → `publish`
- `class CompileKotlin : Task` → `compileKotlin`
- `class ProcessResources : Task` → `processResources`

**Comment by vlsi** [comment]
> Exactly, this is why we can use the class name directly for the task because renaming the class is as much a breaking change as renaming the corresponding task.

What do you think of constants for task names?
I find that the constants are helpful when consuming third-party plugins.

---

I find that "fixed-named" tasks are lifecycle (class-less) in 99.42% of the cases.


```
class Publish : Task → publish
```

`org.gradle.api.publish.plugins.PublishingPlugin#PUBLISH_LIFECYCLE_TASK_NAME`
In other words, `publish` task is classless.

Samples from Apache Calcite:

```

:core:publishCorePublicationToMavenLocal - Publishes Maven publication 'core' to the local Maven repository.
:core:publishCorePublicationToNexusRepository - Publishes Maven publication 'core' to Maven repository 'nexus'.
```

----

```
class ProcessResources : Task → processResources
```

In practice, the task is created for each sourceset, and the name derives from the sourceset;

```
processResources
processTestResources
```

-----

```
class CompileKotlin : Task → compileKotlin
```

```
core:compileKotlin
core:compileTestKotlin
```

The task is per-sourceset, so the default name of `compileKotlin` is not something users need. Every time `compile...Kotlin` is created the task name derives from sourceset.

**Comment by vlsi** [comment]
Just in case, I agree there might be singleton tasks that are created out of classes. However, it looks like the most common case for "singleton tasks" is lifecycle tasks that use `DefaultTask`.

**Comment by Fleshgrinder** [comment]
> What do you think of constants for task names?
> I find that the constants are helpful when consuming third-party plugins.

I have never used such a constant, ever. I use the task group name constants often, but for tasks I have type safe accessors that are automatically generated for me by Gradle. They have auto-completion, and do not require additional imports. What more would I want. 😉 As illustrated in my last comment, only things that are type safe matter, and random constants do not provide type safety, I have no clue what I'm dealing with.

Let me give you other examples without a real task behind that you know, hopefully this helps to dissolve the connection between the real tasks and your knowledge about them:

- `class Foo : Task` → `foo`
- `class GoBar : Task` → `goBar`
- `class GenerateSomeCodeForMe : Task` → `generateSomeCodeForMe`
- `class RunSomethingAwesome : Task` → `runSomethingAwesome`

**Comment by vlsi** [comment]
> I have never used such a constant, ever

Okay :)

The constants for task names are useful when writing plugins as regular classes.
For instance: https://github.com/melix/jmh-gradle-plugin/blob/a30d099ee7c3bd83f8886c84852959d932b1f998/src/main/groovy/me/champeau/jmh/JMHPlugin.groovy#L46

Of course, if you have accessors, they are great. However, if you write plugin as Kotlin or Java or Groovy code, you have no accessors.

> What more would I want.

As a plugin author, you might want to make your plugins accessible for others. 

For instance, here's my plugin that integrates `io.codearte.nexus-staging` and `de.marcphilipp.nexus-publish` for automated publishing:
https://github.com/vlsi/vlsi-release-plugins/blob/83c85c5faa4c7cd1fe0173b75c1cba5e60c3f209/plugins/stage-vote-release-plugin/src/main/kotlin/com/github/vlsi/gradle/release/StageVoteReleasePlugin.kt#L151-L153

```kotlin
// Tasks from NexusStagingPlugin
val closeRepository = tasks.named("closeRepository")
val releaseRepository = tasks.named("releaseRepository")
```

It would be way better if there were constants as follows:

```kotlin
val closeRepository = tasks.named(NexusStagingPlugin.CLOSE_REPOSITORY_TASK_NAME)
val releaseRepository = tasks.named(NexusStagingPlugin.RELEASE_REPOSITORY_TASK_NAME)
```

There will always be cases when Kotlin-generated accessors are not available, so making the API relying on "Gradle-generated accessors for Kotlin" does not sound right.

>hopefully this helps to dissolve the connection between the real tasks and your knowledge about them:

Sorry for nagging you, however, what I learned is that real use-cases do help to better understand the feature request.
It is not enough to suggest that `class Foo` should yield task `foo`.
You need to clarify what problem are you trying to solve with the request. Of course, you might ignore that altogether and wait for the Gradle team to pick it up and analyze, however, there's enough stuff on the plate, so I would suggest coming up with the real case for the demand rather than suggesting the feature "because it can be implemented".

**Comment by Fleshgrinder** [comment]
```kotlin
val closeRepository by tasks.existing
val releaseRepository by tasks.existing
```

Why wouldn't you use this syntax? Sure, in Java and Groovy you cannot use it, and there the constants are surely useful. But most plugins I develop are in-house and everything is Kotlin only.

> Sorry for nagging you, however, what I learned is that real use-cases do help to better understand the feature request.

DRY, it's so simple and I struggle to give you more examples, it's all about DRY, DRY, and DRY. I want to get rid of code like this:

```kotlin
val task = project.tasks.register<GoFoo>("goFoo")
val task = project.tasks.register<GoFoo>(GoFoo.GO_FOO_TASK_NAME)
```

Note that this is plugin internal code, not user code, user's have other means to get the tasks they want:

```kotlin
tasks.goFoo {}
val goFoo by tasks.existing
```

Also, if this part of the request is so controversial to you then we probably should move it to a dedicated issue to not block the rest.

**Comment by vlsi** [comment]
>val closeRepository by tasks.existing
> val releaseRepository by tasks.existing
> Why wouldn't you use this syntax?

It might be something with stale Gradle or `kotlin-dsl` version.
For some reason `by` delegates did not work, and I did not investigate much.

However, delegates there do not make it easier to understand where the task comes from.

**Comment by vlsi** [comment]
>Also, if this part of the request is so controversial to you then we probably should move it to a dedicated issue to not block the rest.

Please do so. Inferring the task names from class names is extremely controversal

**Comment by sschuberth** [comment]
> For `registering` it would be nice to have the ability, as we already have it for `register`, to pass the type via a type parameter, instead of a value parameter.

Absolutely!

> Of course I'm willing to provide the implementation with tests in a PR if this change is deemed useful.

Personally, I'd appreciate a PR for even just that!

**Comment by ov7a** [comment]
Sorry for the late reply

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by jbartok** [comment]
If (big one) we were to have the feature of deriving task names from task types, then it should be a feature available in the Gradle API, i.e., not specific to the Kotlin DSL. 

Cortex team, how do you feel about having this feature?

**Comment by mlopatkin** [comment]
> If (big one) we were to have the feature of deriving task names from task types, then it should be a feature available in the Gradle API, i.e., not specific to the Kotlin DSL.
> 
> Cortex team, how do you feel about having this feature?

We discussed it and we don't want to introduce automatic type-based task names. It is a nice convention to adopt, but it isn't a best practice Gradle endorses today. Those who want it are encouraged to implement a helper function themselves.

A reified overload for `registering` would be nice, IMO, and would give a nice symmetry with `register`. This is, however, something for the Kotlin DSL maintainers to decide upon. There is also another issue about the matter:
* #21738

I suggest closing this issue and merging the `registering<FooBar>` use case into the issue above.

**Comment by eskatos** [comment]
Thank you @mlopatkin for clarifying your position on inferred task names. I also agree with your assessment that this should be combined with the other issue.

I'm closing this as a duplicate of
* https://github.com/gradle/gradle/issues/21738

@Fleshgrinder, @vlsi, please add the `registering<FooBar>` use case into the issue above.
